### PR TITLE
Clean-up ocean wetting and drying routine

### PR DESCRIPTION
**Pull request by cbegeman**
Clean-up the wetting and drying routine in MPAS-Ocean and make a few bugfixes in that routine. 

This PR utilizes a new variable `wettingVelocityFactor`, which replaces `wettingVelocity`, to scale both `normalVelocity` and `normalVelocityTend` in instances where the `layerThickness` is approaching a minimum `layerThickness`. 

We preserve the existing option where `config_zero_drying_velocity = True`, zeroing out `normalVelocity` and `normalVelocityTend` using `wettingVelocityFactor` in regions where the divergence out of a cell would bring the `layerThickness` within a tolerance of the minimum thickness.

We remove an approach for setting `wettingVelocityFactor` between 0 and 1 when `config_zero_drying_velocity = False` because this code was never operational. Instead when `config_zero_drying_velocity = False` there is no velocity or tendency modification in drying cells. 

Bugfix 1 is changing the time step used in the wetting and drying routine to the RK4 substep. Bugfix 2 is changing range over which `wettingVelocityFactor` is computed to active connected edges.

[BFB]

### CODE REVIEW

**Comment by cbegeman** [comment]
This PR is migrated from https://github.com/E3SM-Ocean-Discussion/E3SM/pull/14. 

This commit was added since that review [26f525c](https://github.com/E3SM-Project/E3SM/pull/5418/commits/26f525c3e351b4372ee237161f406fb35dbbd59d). With this commit, wetting and drying tests are BFB compared to the version without this commit.

**Comment by xylar** [comment]
@sbrus89, would you be willing to do some tests to make sure things still behave as expected, following up on the testing we did in https://github.com/E3SM-Ocean-Discussion/E3SM/pull/14?

I'll do some "do no harm" testing in compass to make sure we're good there.

@jonbob, would you be up for running a couple of relevant E3SM tests to make sure we're BFB as expected.  Or feel free to recommend some tests for me to run instead.

**Comment by sbrus89** [comment]
Yes, I'll do that.

**Review by xylar — APPROVED** [review]
I looked over the code and it looks fine to me.  My review is mostly from a do-no-harm perspective, as I am not super well versed on the details of the wetting-and-drying code.

The compass `pr` test suite was BFB on Chrysalis with Intel and OpenMPI.

**Comment by rljacob** [comment]
@sbrus89 please review.

**Comment by cbegeman** [comment]
Through conversations with @sbrus89, we realized that the change from `rk_substep_weight` to `dt` https://github.com/E3SM-Project/E3SM/pull/5418/commits/4329f78bd8514ce865fc291078b7504a50b0b04f is best thought of not as a bugfix but rather a change that makes the wetting and drying routine more conservative (more likely to use `wettingVelocityFactor` as a limiter on the velocity).

In my `drying_slope` and `dam_break` tests, I did not find that https://github.com/E3SM-Project/E3SM/pull/5418/commits/4329f78bd8514ce865fc291078b7504a50b0b04f made a difference to the solution. @sbrus89 is exploring whether dropping this commit makes a difference for resolution convergence in his new parabolic bowl test with wetting and drying. 

We also discussed dropping this line in the code, https://github.com/E3SM-Project/E3SM/blob/26f525c3e351b4372ee237161f406fb35dbbd59d/components/mpas-ocean/src/shared/mpas_ocn_wetting_drying.F#L366 in an attempt to improve the parabolic bowl convergence, which would also make the wetting and drying routine less conservative.

@sbrus89 is planning on reporting back on the outcome of these tests before approving the PR. Please let me know if I represented our conversation accurately!

**Comment by rljacob** [comment]
@sbrus89 still needs your input.

**Comment by sbrus89** [comment]
Thanks for the reminder, @rljacob. I'm working on testing now.

**Comment by sbrus89** [comment]
@cbegeman, After rebasing locally on E3SM/master (0f8188) and testing inside compass (3a4bef) on compy, the drying slope cases run and give good results. However, I'm seeing failures with both dam break cases.

**Comment by cbegeman** [comment]
@sbrus89 Did the dam break case pass on master? I just ran with E3SM/master at 09498b6efc and that fails on chrysalis, intel, openmpi with validate state errors.

**Comment by sbrus89** [comment]
@cbegeman - I did not check, but yes I would suspect that is the case. I can run compass bisect to track down the commit that introduced the errors.

**Comment by cbegeman** [comment]
@sbrus89 It must have been relatively recently. I don't recall how long ago I tested dam break for this PR (https://github.com/E3SM-Ocean-Discussion/E3SM/pull/14 was around for a while), but the dam break case was definitely working when we were testing https://github.com/E3SM-Project/E3SM/pull/5279.

**Comment by sbrus89** [comment]
I agree, it couldn't have been from too long ago.

**Comment by cbegeman** [comment]
@sbrus89 If you have time and interest in running git bisect that would be much appreciated. Let me know if you want to pass this off to me.

**Comment by sbrus89** [comment]
sure, I'll get it set up today and let you know how it goes.

**Comment by sbrus89** [comment]
@cbegeman - Here's what I found out after running the compass bisect utility:

```
627683e64a5d01e24c1468d589bbcfa98a0c30e2 is the first bad commit
commit 627683e64a5d01e24c1468d589bbcfa98a0c30e2
Merge: 146732cb27 d0b11efec9
Author: Jon Wolfe <jonbob@lanl.gov>
Date:   Mon Dec 5 14:34:11 2022 -0600

    Merge branch 'cbegeman/ocn/mask-time-series-stats-vars' (PR #5325)
    
    Mask ocean time series stats variables
    
    This PR applies fill value masking to non-restart time series stats
    streams in the ocean component. It also specifies the masking array for
    a subset of variables in the Registry.
    
    Fixes #5253
    
    [BFB]
```

**Comment by xylar** [comment]
Man, lesson learned!  If you don't do proper CF masking from the beginning, you really will have trouble adding it later. Hopefully, we can build this into Omega from the ground up!

**Comment by cbegeman** [comment]
@sbrus89 Thanks for the testing. Why am I not surprised it was that PR! I'm figuring out the minimum changes needed for dam break to pass. I should have an answer soon.

**Comment by cbegeman** [comment]
@sbrus89 You can run the dam_break test case and verify that the solution still agrees just as well with the data after this PR. This is the commit that you need https://github.com/cbegeman/E3SM/commit/39192f6c6607512482dbea7e4cdcc4b28f29fe6f. I don't like this solution though, we might need some change in how tracers are treated in cases where their tendencies are disabled. I'd like to open a separate issue documenting this problem and address it separately from this PR, since the issue isn't directly associated with this PR. Does that sound good to you?

**Comment by sbrus89** [comment]
Sure, no problem. I'll report back.

**Comment by cbegeman** [comment]
@sbrus89 I posted a PR here https://github.com/E3SM-Project/E3SM/pull/5519 that provides the bugfix without needing to revert fill-value masking.

**Comment by sbrus89** [comment]
@cbegeman, great! Reverting the fill value worked, but I'll test with your new PR instead. Seems like that's a better option.

**Comment by sbrus89** [comment]
I verified this works with #5519 for the dam break.

File: components/mpas-ocean/src/shared/mpas_ocn_wetting_drying.F [hunk 1]
```diff
@@ -371,73 +348,49 @@ subroutine ocn_wetting_drying_wettingVelocity(meshPool, layerThickEdgeFlux, laye
...
+                divOutFlux = divOutFlux &
+                             + normalVelocity(k, iEdge) * edgeSignOnCell(i, iCell) * &
+                               layerThickEdgeFlux(k, iEdge) * dvEdge(iEdge) * &
+                               invAreaCell(iCell)
               end if
             end if
           end do
 
           ! if layer thickness is too small, limit divergence flux outwards with opposite velocity
-          if ((layerThickness + dt*divOutFlux ) <= (config_drying_min_cell_height + config_drying_safety_height))  then
```
[the remark this answers is not in the document]
Comment (cbegeman) [reply components/mpas-ocean/src/shared/mpas_ocn_wetting_drying.F:422]: @sbrus89 Thanks for the reminder. I dropped the commit that made the change https://github.com/E3SM-Project/E3SM/pull/5418/commits/4329f78bd8514ce865fc291078b7504a50b0b04f, which modified the `dt` input to the subroutine. The tests pass the nightly suite and there are machine-precision differences for the drying_slope cases.

**Review by sbrus89 — APPROVED** [review]
@cbegeman, this is great. thanks!

**Comment by cbegeman** [comment]
@jonbob This should be ready for E3SM testing. I think that this doesn't fit the definition of a stealth PR: "BFB code change that includes a new feature, turned off by default" because this PR does not introduce the stealth feature (wetting and drying). However, I'm uncertain about the intention of "includes a new feature" here.

**Comment by jonbob** [comment]
@cbegeman -- at this point E3SM doesn't even have the wetting and drying namelist, so I think it's safe to bring is as something that only impacts the standalone model

**Comment by jonbob** [comment]
passes:
* SMS_D_Ld3.T62_oQU120.CMPASO-IAF.chrysalis_intel
* SMS_D_Ld1.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel.allactive-wcprod
* ERS.ne11_oQU240.WCYCL1850NS.chrysalis_intel

merged to next

**Comment by cbegeman** [comment]
@jonbob Thanks for the testing!

**Pull request merged**